### ISSUE TITLE
Implement a unified (single) GNUmakefile for each simulation (machine)

### ISSUE THREAD

**Issue by dmcnaugh** [udo-munk/z80pack#53]
### Motivation
- I have felt sorry about all the `Makefile.*` modifications my new code has caused
- I want to learn more about `make` as I have only used it very simply in the past

### Goals:
- have a single Makefile that works for all platforms for a given simulation eg. **imsaisim**
- localise the OS specific variations into a single place 
- **not** have to symlink from `../../z80sim/simsrc`
- **not** have to use `#include "../../something/more.h"` to reference header files in the source
- allow `make install` to well defined paths
- have best-practice file dependancy (there are some errors here and there at the moment)
- make it very simple to list the source files that are required to build a given machine

feel free to make other suggestions...

**Comment by udo-munk** [comment]
For creating a single Makefile you need to test on all supported platforms.
OSX, BSD, Linux and Cygwin is good enough. I've used ports to SPARC, IBM Power etc. only for finding problems related to byte order and for performance records. Ususally people don't have such systems at home, no need for a build system and someone using the platforms can build anyway.

You cannot assume that a platform uses GNU make, just because OSX does. Technically one could install GNU make on such platforms, but of course another make already exists, and this will require quite some workarrounds.

File dependancy can automatically be created with make depend and scripting. This stuff will rewite the Makefile it self of course and it can screw up. So far I have avoided this, because the project is small enough to do this manually. The missing dependancies are not really errors, but lazyness ;-)

When writing Makefiles keep in mind that every user is supposed to work with it. There are no precompiled packages, I won't build them, is the job of package managers anyway. And then, IMHO, this is a computer construction kit for building stuff self, so users should do this and be enabled to do so. An over complex build system only manageable by experienced software developers might not be helpfull.

**Comment by udo-munk** [comment]
For testing with other platforms I'm using Virtualbox and QEMU, both can be installed on OSX, Linux and Windows at least. Using throwaway VM's with some OS installation is the fastest way for a test build, but still time consuming, but needs to be done from time to time.

Thanks, such errors should be fixed of course.

I know and I appreciate your help, thanks.

**Comment by udo-munk** [comment]
I've updated the old makefiles so that all machines can be build.
Then I added GNUmakfile to the Altair to see if I get this straight.

Looks good so far, good work. Have to test the GNUmakefiles on some other platforms to see how it works there and if changes are needed.

**Comment by udo-munk** [comment]
Testing on Debian Linux, the TARGET_OS stuff isn't working correct:

udo@debian:~/work/cpm/z80pack/imsaisim/srcsim$ make
make: *** No rule to make target '/opt/X11/include/X11/Xlib.h', needed by 'simctl.o'.  Stop.

I still uses the plaform includes for OSX.

**Comment by dmcnaugh** [comment]
OK. I had a chance to test this on Ubuntu 16.04 (simply because it was as image I already had available). 
I didn't get the same error as you, but I found there was a problem with the ordering of the libraries in the `LFLAGS` definition.
```
LFLAGS = $(PLAT_LFLAGS) \
	-ljpeg -lGL -lGLU -lX11 \
        -L$(DIR_FP) -L$(DIR_CIV) $(SRC_LIBS)
```
needs to be modified to read
```
LFLAGS = -L$(DIR_FP) -L$(DIR_CIV) $(SRC_LIBS) \
        -ljpeg -lGL -lGLU -lX11 \
        $(PLAT_LFLAGS)
```
With this simple change I was able to build on this platform (and it still works on OSX).

One thing that has caused me trouble many times while testing  is forgetting to clean-up between builds using the 2 different make files. If you make with `make -f Makefile.os_plat` then always `make -f Makefile.os_plat clean` before trying `make` with the GNUmakefile, and vice-versa.
